### ISSUE TITLE
Curse of float64

### ISSUE THREAD

**Issue by unnonouno** [pfnet/chainer#139]
Adding `int` to 0-dim array causes a `float64` value.

```
>>> (numpy.array(1, numpy.float32) + 1).dtype
dtype('float64')
```

When its dimension is greater than 0, this problem isn't happen.
```
>>> (numpy.array([1], numpy.float32) + 1).dtype
dtype('float32’)
```

Only functions in `basic_math` accept 0-dim array as an input value.
These functions only forward input data to numpy, and it may causes `float64` problem.

We need to check types of arguments and when they contains `float32` and standard python types like `int` or `float`, we need to convert results to `float32` as results of calculation are `float64`.

For example, this code is dangerous:
```
loss = softmax_cross_entropy(...) / 10
loss.backward()
```

**Comment by unnonouno** [comment]
```
>>> x = chainer.Variable(numpy.array(1, numpy.float32))
>>> (x + 1).data.dtype
dtype('float64')
```
